KOKKOS CMAKE issue

@junghans I'm trying to build KOKKOS-enabled LAMMPS "23 Oct 2017" using CMAKE but I'm getting an error:
```
[ 59%] Building CXX object CMakeFiles/lammps.dir/home/stamoor/Documents/lammps_master/src/neighbor.cpp.o
[ 59%] Building CXX object CMakeFiles/lammps.dir/home/stamoor/Documents/lammps_master/src/set.cpp.o
[ 59%] Building CXX object CMakeFiles/lammps.dir/home/stamoor/Documents/lammps_master/src/ntopo_angle_all.cpp.o
lammps_master/cmake/../src/KOKKOS/npair_copy_kokkos.h: In member function 'void LAMMPS_NS::Neighbor::init_styles()':
lammps_master/cmake/../src/KOKKOS/npair_copy_kokkos.h:17:28: error: 'LMPDeviceType' was not declared in this scope
            NPairCopyKokkos<LMPDeviceType>,
                            ^
lammps_master/src/neighbor.cpp:586:38: note: in definition of macro 'NPairStyle'
   pairclass[npclass] = &pair_creator<Class>; \
                                      ^
lammps_master/cmake/../src/KOKKOS/npair_copy_kokkos.h:17:41: error: template argument 1 is invalid
            NPairCopyKokkos<LMPDeviceType>,
```
Any ideas?

I also had to manually set the environment variable KOKKOS_PATH on the make line, we may want to add a note for this to the Readme.

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -481,9 +481,7 @@ if(ENABLE_KOKKOS)
                          ${KOKKOS_PKG_SOURCES_DIR}/neigh_bond_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/fix_nh_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/nbin_kokkos.cpp
-                         ${KOKKOS_PKG_SOURCES_DIR}/nbin_ssa_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/npair_kokkos.cpp
-                         ${KOKKOS_PKG_SOURCES_DIR}/npair_ssa_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/domain_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/modify_kokkos.cpp)
   set_property(GLOBAL PROPERTY "KOKKOS_PKG_SOURCES" "${KOKKOS_PKG_SOURCES}")
@@ -492,8 +490,15 @@ if(ENABLE_KOKKOS)
   RegisterStylesExt(${KOKKOS_PKG_SOURCES_DIR} kokkos KOKKOS_PKG_SOURCES)
 
   # register kokkos-only styles
-  RegisterNBinStyles(${KOKKOS_PKG_SOURCES_DIR})
-  RegisterNPairStyles(${KOKKOS_PKG_SOURCES_DIR})
+  RegisterNBinStyle(${KOKKOS_PKG_SOURCES_DIR}/nbin_kokkos.h)
+  RegisterNPairStyle(${KOKKOS_PKG_SOURCES_DIR}/npair_kokkos.h)
+
+  if(ENABLE_USER-DPD)
+    get_property(KOKKOS_PKG_SOURCES GLOBAL PROPERTY KOKKOS_PKG_SOURCES)
+    list(APPEND KOKKOS_PKG_SOURCES ${KOKKOS_PKG_SOURCES_DIR}/npair_ssa_kokkos.cpp)
+    RegisterNPairStyle(${KOKKOS_PKG_SOURCES_DIR}/npair_ssa_kokkos.h)
+    set_property(GLOBAL PROPERTY "KOKKOS_PKG_SOURCES" "${KOKKOS_PKG_SOURCES}")
+  endif()
 
   get_property(KOKKOS_PKG_SOURCES GLOBAL PROPERTY KOKKOS_PKG_SOURCES)
 

--- a/cmake/CMakeLists.txt
+++ b/cmake/CMakeLists.txt
@@ -480,12 +480,20 @@ if(ENABLE_KOKKOS)
                          ${KOKKOS_PKG_SOURCES_DIR}/neigh_list_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/neigh_bond_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/fix_nh_kokkos.cpp
+                         ${KOKKOS_PKG_SOURCES_DIR}/nbin_kokkos.cpp
+                         ${KOKKOS_PKG_SOURCES_DIR}/nbin_ssa_kokkos.cpp
+                         ${KOKKOS_PKG_SOURCES_DIR}/npair_kokkos.cpp
+                         ${KOKKOS_PKG_SOURCES_DIR}/npair_ssa_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/domain_kokkos.cpp
                          ${KOKKOS_PKG_SOURCES_DIR}/modify_kokkos.cpp)
   set_property(GLOBAL PROPERTY "KOKKOS_PKG_SOURCES" "${KOKKOS_PKG_SOURCES}")
 
   # detects styles which have KOKKOS version
   RegisterStylesExt(${KOKKOS_PKG_SOURCES_DIR} kokkos KOKKOS_PKG_SOURCES)
+
+  # register kokkos-only styles
+  RegisterNBinStyles(${KOKKOS_PKG_SOURCES_DIR})
+  RegisterNPairStyles(${KOKKOS_PKG_SOURCES_DIR})
 
   get_property(KOKKOS_PKG_SOURCES GLOBAL PROPERTY KOKKOS_PKG_SOURCES)
 

--- a/cmake/Modules/StyleHeaderUtils.cmake
+++ b/cmake/Modules/StyleHeaderUtils.cmake
@@ -11,6 +11,12 @@ function(FindStyleHeaders path style_class file_pattern headers)
     set_property(GLOBAL PROPERTY ${headers} "${hlist}")
 endfunction(FindStyleHeaders)
 
+function(AddStyleHeader path headers)
+    get_property(hlist GLOBAL PROPERTY ${headers})
+    list(APPEND hlist ${path})
+    set_property(GLOBAL PROPERTY ${headers} "${hlist}")
+endfunction(AddStyleHeader)
+
 function(FindStyleHeadersExt path style_class extension headers sources)
     get_property(hlist GLOBAL PROPERTY ${headers})
     get_property(slist GLOBAL PROPERTY ${sources})
@@ -69,6 +75,14 @@ endfunction(RegisterNBinStyles)
 function(RegisterNPairStyles search_path)
     FindStyleHeaders(${search_path} NPAIR_CLASS     npair_     NPAIR     ) # npair     ) # neighbor
 endfunction(RegisterNPairStyles)
+
+function(RegisterNBinStyle path)
+    AddStyleHeader(${path} NBIN)
+endfunction(RegisterNBinStyle)
+
+function(RegisterNPairStyle path)
+    AddStyleHeader(${path} NPAIR)
+endfunction(RegisterNPairStyle)
 
 function(RegisterStyles search_path)
     FindStyleHeaders(${search_path} ANGLE_CLASS     angle_     ANGLE     ) # angle     ) # force

--- a/cmake/Modules/StyleHeaderUtils.cmake
+++ b/cmake/Modules/StyleHeaderUtils.cmake
@@ -62,6 +62,14 @@ function(GenerateStyleHeader path property style)
     CreateStyleHeader("${path}" "style_${style}.h" ${files})
 endfunction(GenerateStyleHeader)
 
+function(RegisterNBinStyles search_path)
+    FindStyleHeaders(${search_path} NBIN_CLASS      nbin_      NBIN      ) # nbin      ) # neighbor
+endfunction(RegisterNBinStyles)
+
+function(RegisterNPairStyles search_path)
+    FindStyleHeaders(${search_path} NPAIR_CLASS     npair_     NPAIR     ) # npair     ) # neighbor
+endfunction(RegisterNPairStyles)
+
 function(RegisterStyles search_path)
     FindStyleHeaders(${search_path} ANGLE_CLASS     angle_     ANGLE     ) # angle     ) # force
     FindStyleHeaders(${search_path} ATOM_CLASS      atom_vec_  ATOM_VEC  ) # atom      ) # atom      atom_vec_hybrid

